pkg/meta/types.go

```diff
@@ -0,0 +1,129 @@
+/*
+ * JuiceFS, Copyright (C) 2021 Juicedata, Inc.
+ *
+ * This program is free software: you can use, redistribute, and/or modify
+ * it under the terms of the GNU Affero General Public License, version 3
+ * or later ("AGPL"), as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package meta
+
+type DumpedCounters struct {
+	UsedSpace         int64 `json:"usedSpace"`
+	UsedInodes        int64 `json:"usedInodes"`
+	NextInode         int64 `json:"nextInodes"`
+	NextChunk         int64 `json:"nextChunk"`
+	NextSession       int64 `json:"nextSession"`
+	NextCleanupSlices int64 `json:"nextCleanupSlices"`
+}
+
+type DumpedDelFile struct {
+	Inode  Ino    `json:"inode"`
+	Length uint64 `json:"length"`
+	Expire int64  `json:"expire"`
+}
+
+type DumpedSustained struct {
+	Sid    uint64 `json:"sid"`
+	Inodes []Ino  `json:"inodes"`
+}
+
+type DumpedAttr struct {
+	Type      string `json:"type"`
+	Mode      uint16 `json:"mode"`
+	Uid       uint32 `json:"uid"`
+	Gid       uint32 `json:"gid"`
+	Atime     int64  `json:"atime"`
+	Mtime     int64  `json:"mtime"`
+	Ctime     int64  `json:"ctime"`
+	Atimensec uint32 `json:"atimensec"`
+	Mtimensec uint32 `json:"mtimensec"`
+	Ctimensec uint32 `json:"ctimensec"`
+	Nlink     uint32 `json:"nlink"`
+	Length    uint64 `json:"length,omitempty"`
```

Comment:
0 length of a file should not be omitted

Reply:
b936bea